src/router.js

```diff
@@ -56,6 +56,10 @@ export default function router(history, routes) {
       if (match) {
         lastMatch = match;
         effects.push(spawn(match.action, match.params));
+
+        if ('beforeRouteChange' in options) {
+          effects.unshift(spawn(options.beforeRouteChange, match.params));
```

Comment:
By placing this in the array, it will potentially run in parallel to the matched route instead of technically running before it. We'll need to ensure that it can run first. If you need any help with this, let me know.